src/libsystemd/sd-journal/journal-verify.c

```diff
@@ -1135,19 +1139,37 @@ int journal_file_verify(
 
 #if HAVE_GCRYPT
                         if (JOURNAL_HEADER_SEALED(f->header)) {
-                                uint64_t q, rt;
+                                uint64_t q, rt, rt_end;
 
                                 debug(p, "Checking tag %"PRIu64"...", le64toh(o->tag.seqnum));
 
                                 rt = f->fss_start_usec + le64toh(o->tag.epoch) * f->fss_interval_usec;
-                                if (entry_realtime_set && entry_realtime >= rt + f->fss_interval_usec) {
+                                rt_end = usec_add(rt, f->fss_interval_usec);
+                                if (entry_realtime_set && entry_realtime >= rt_end) {
                                         error(p,
                                               "tag/entry realtime timestamp out of synchronization (%"PRIu64" >= %"PRIu64")",
                                               entry_realtime,
                                               rt + f->fss_interval_usec);
                                         r = -EBADMSG;
                                         goto fail;
                                 }
+                                if (max_entry_realtime >= rt_end) {
+                                        error(p,
+                                              "Entry realtime (%"PRIu64", %s) is too late with respect to tag (%"PRIu64", %s)",
+                                              max_entry_realtime, FORMAT_TIMESTAMP(max_entry_realtime),
+                                              rt_end, FORMAT_TIMESTAMP(rt_end));
+                                        r = -EBADMSG;
+                                        goto fail;
+                                }
+                                if (min_entry_realtime < rt) {
+                                        error(p,
+                                              "Entry realtime (%"PRIu64", %s) is too early with respect to tag (%"PRIu64", %s)",
+                                              min_entry_realtime, FORMAT_TIMESTAMP(min_entry_realtime),
+                                              rt, FORMAT_TIMESTAMP(rt));
+                                        r = -EBADMSG;
+                                        goto fail;
+                                }
+                                min_entry_realtime = USEC_INFINITY;
```

Comment:
Why is this re-initialized here? If this is necessary, why not for `max_entry_realitime` ?

Reply:
The Journal is a sequence of (usually time-increasing) entries, interleaved with guaranteed increasing tag objects. The tag objects can authenticate entries in non-overlapping time intervals. After one tag object is processed (and verified that all entries it authenticates fall withing its time interval), the verification of the next tag starts. As the next tag will not be able to authenticate the messages from the current tag, we need to reset `min_entry_realtime`. We could also reset `max_entry_realtime`, but as the epochs are increasing it doesn't make a difference. In fact, the verification of the next `min_entry_realtime` ensures that resetting `max_entry_realtime` makes no difference. So I don't mind and we can reset it or not, whatever you like more.

Reply:
Indeed. Thank you for the explanation.